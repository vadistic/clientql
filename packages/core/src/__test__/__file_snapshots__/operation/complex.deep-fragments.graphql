query FindEventsAtVenueMyQuery($venueId: ID!) {
  findEventsAtVenue(venueId: $venueId) {
    ...EventDeep
  }
}

fragment EventDeep on Event {
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  ... on Concert {
    ...ConcertDeep
  }
  ... on Festival {
    ...FestivalDeep
  }
  ... on Conference {
    ...ConferenceDeep
  }
}

fragment VenueDeep on Venue {
  __typename
  id
  name
  address
  maxOccupancy
}

fragment ConcertDeep on Concert {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  previousVenues {
    ...VenueDeep
  }
  minAgeRestriction
  performingBand {
    ...PerformerDeep
  }
}

fragment PerformerDeep on Performer {
  __typename
  id
  name
  contact
  speciality
  fee
}

fragment FestivalDeep on Festival {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  performers {
    ...PerformerDeep
  }
}

fragment ConferenceDeep on Conference {
  __typename
  id
  name
  startsAt
  endsAt
  venue {
    ...VenueDeep
  }
  minAgeRestriction
  speakers {
    ...SpeakerDeep
  }
  workshops
}

fragment SpeakerDeep on Speaker {
  __typename
  id
  name
  contact
}
