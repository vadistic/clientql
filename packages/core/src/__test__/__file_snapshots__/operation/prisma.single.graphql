query UserQuery($where: UserWhereUniqueInput!) {
  user(where: $where) {
    ...UserFlatPartial
    posts {
      ...PostFlatPartial
    }
  }
}

fragment UserFlatPartial on User {
  __typename
  id
  nationality
  email
  name
}

fragment PostFlatPartial on Post {
  __typename
  id
  content
  createdAt
  published
  title
  updatedAt
}
