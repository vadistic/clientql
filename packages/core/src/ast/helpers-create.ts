import {
  ArgumentNode,
  FieldNode,
  FragmentDefinitionNode,
  FragmentSpreadNode,
  InlineFragmentNode,
  InputValueDefinitionNode,
  NameNode,
  OperationDefinitionNode,
  OperationTypeNode,
  SelectionNode,
  VariableDefinitionNode,
} from 'graphql'
import { Kind } from './kind'
import { Typename } from './types'

export const createName = (name: string): NameNode => ({
  kind: Kind.NAME,
  value: name,
})

export interface CreateFieldProps {
  fieldname: string
  arguments?: ArgumentNode[]
  selections?: SelectionNode[]
}

export const createField = (props: CreateFieldProps): FieldNode => ({
  kind: Kind.FIELD,
  name: createName(props.fieldname),
  arguments: props.arguments,
  selectionSet: props.selections
    ? {
        kind: Kind.SELECTION_SET,
        selections: props.selections,
      }
    : undefined,
})

export interface CreateFragmentProps {
  fragmentname: string
  condition: Typename
  selections: SelectionNode[]
}
/**
 *  no fragment variables (yet?)
 */
export const createFragment = ({
  condition,
  selections,
  fragmentname,
}: CreateFragmentProps): FragmentDefinitionNode => ({
  kind: Kind.FRAGMENT_DEFINITION,
  name: createName(fragmentname),
  typeCondition: {
    kind: Kind.NAMED_TYPE,
    name: createName(condition),
  },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections,
  },
})

export const createFragmentSpread = (name: string): FragmentSpreadNode => ({
  kind: Kind.FRAGMENT_SPREAD,
  name: createName(name),
})

export const createInlineFragment = (
  on: Typename,
  selections: SelectionNode[],
): InlineFragmentNode => ({
  kind: Kind.INLINE_FRAGMENT,
  typeCondition: {
    kind: Kind.NAMED_TYPE,
    name: createName(on),
  },
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections,
  },
})

export const createOperationVariable = (suffix: string) => (
  node: InputValueDefinitionNode,
): VariableDefinitionNode => ({
  kind: Kind.VARIABLE_DEFINITION,
  defaultValue: node.defaultValue,
  directives: node.directives,
  variable: {
    kind: Kind.VARIABLE,
    name: createName(node.name.value + suffix),
  },
  type: node.type,
})

export const createOperationArgument = (suffix: string) => (
  node: InputValueDefinitionNode,
): ArgumentNode => ({
  kind: Kind.ARGUMENT,
  name: node.name,
  value: {
    kind: Kind.VARIABLE,
    name: createName(node.name.value + suffix),
  },
})

export interface CreateOperationProps {
  name: string
  type: OperationTypeNode
  variables?: VariableDefinitionNode[]
  selections: SelectionNode[]
}

export const createOperation = (props: CreateOperationProps): OperationDefinitionNode => ({
  kind: Kind.OPERATION_DEFINITION,
  name: createName(props.name),
  operation: props.type,
  selectionSet: {
    kind: Kind.SELECTION_SET,
    selections: props.selections,
  },
  variableDefinitions: props.variables,
})
