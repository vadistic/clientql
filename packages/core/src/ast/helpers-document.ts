import { ASTKindToNode, DefinitionNode, DocumentNode, KindEnum, SelectionNode } from 'graphql'
import { truthy } from '../utils/types'
import { Kind } from './kind'

/**
 * created doc node fromd efinition/s
 */

export const wrapDocument = (
  ...nodes: Array<DefinitionNode | undefined | null | false>
): DocumentNode => ({
  kind: Kind.DOCUMENT,
  definitions: nodes.filter(truthy),
})

/**
 * get mutable arr of definitions from doc node
 */

export const unwrapDocument = (doc: DocumentNode): DefinitionNode[] => [...doc.definitions]

/**
 * get first/firstOfKind definition from docuemnt
 */
export interface GetDocDefinition {
  <K extends KindEnum>(doc: DocumentNode, kind: K): ASTKindToNode[K] | undefined
  (doc: DocumentNode): DefinitionNode
}

export const getDocDefinition: GetDocDefinition = <K extends KindEnum>(
  doc: DocumentNode,
  kind?: K,
) => {
  if (!doc.kind || doc.kind !== Kind.DOCUMENT || !Array.isArray(doc.definitions)) {
    throw Error('getDocDefinition: Invalid Document (like no document at all')
  }

  if (kind) {
    for (const node of doc.definitions) {
      if (node.kind === kind) {
        return node
      }
    }

    // this should be automatic...
    return undefined as any
  }

  return doc.definitions[0]
}

// for interfaces/unions that need it flat
export const unwrapSelectionSet = (node?: SelectionNode) =>
  (node && node.kind === Kind.FIELD && node.selectionSet && [...node.selectionSet.selections]) ||
  undefined
