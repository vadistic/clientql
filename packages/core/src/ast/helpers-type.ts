import { ListTypeNode, NamedTypeNode, TypeNode } from 'graphql'
import { Kind, KindMap } from './kind'

/**
 * alternative to those nested Null/List types that is easier to print
 */
export type TypeModifier = KindMap['NON_NULL_TYPE'] | KindMap['LIST_TYPE']

export interface UnwrapType {
  type: NamedTypeNode
  typename: string
  modifiers: TypeModifier[]
}

export const unwrapType = (
  node: TypeNode,
  modifiers: TypeModifier[] = [],
): UnwrapType => {
  if (node.kind === Kind.NON_NULL_TYPE) {
    return unwrapType(node.type, [...modifiers, Kind.NON_NULL_TYPE])
  }

  if (node.kind === Kind.LIST_TYPE) {
    return unwrapType(node.type, [...modifiers, Kind.LIST_TYPE])
  }

  // node.kind === Kind.NAMED_TYPE
  return {
    type: node,
    typename: node.name.value,
    modifiers,
  }
}

/**
 * same as unwarpType but without those modifiers when not needed
 */

export const getTypename = (node: TypeNode): string =>
  node.kind === Kind.NAMED_TYPE ? node.name.value : getTypename(node.type)

/**
 * means top level value is nullable
 */
export const isNullable = (
  node: TypeNode,
): node is ListTypeNode | NamedTypeNode => node.kind !== Kind.NON_NULL_TYPE

/**
 * recursively check if field type will be a list
 */

export const isList = (node: TypeNode): boolean =>
  node.kind === Kind.LIST_TYPE
    ? true
    : node.kind === Kind.NAMED_TYPE
    ? false
    : isList(node.type)
