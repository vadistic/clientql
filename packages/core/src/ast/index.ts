export * from './helpers-type'
export * from './helpers-document'
export * from './helpers-create'

export * from './kind'
export * from './merge-extensions'
export * from './guards'
export * from './types'
