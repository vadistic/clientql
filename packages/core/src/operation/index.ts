export * from './build-selections'
export * from './build-operation'
export * from './utils'
export * from './types'
