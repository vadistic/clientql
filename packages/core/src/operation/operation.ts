import {
  ArgumentNode,
  FieldDefinitionNode,
  FieldNode,
  FragmentDefinitionNode,
  OperationTypeNode,
  SelectionNode,
  VariableDefinitionNode,
} from 'graphql'
import {
  createField,
  createOperation,
  createOperationArgument,
  createOperationVariable,
  Fieldname,
  Typename,
  wrapDocument,
} from '../ast'
import { CoreProps } from '../config'
import { buildSelection, retriveCacheFragments } from './selection'

/**
 * this needs to build operation from arr of fieldnames
 */

export const buildOperation = (props: CoreProps) => (path: Fieldname[]) => {
  const [head, ...tail] = path
  const { type, vtx } = determineOperationType(props)(head)

  // head not on any of roots
  if (!vtx || !type) {
    return
  }

  const { fragments, selections, variables } = buildDeepOperation(props)(
    vtx.name,
    path,
  )

  const operation = createOperation({
    name: 'TODO',
    selections,
    type,
    variables,
  })

  return wrapDocument(operation, ...fragments)
}

interface BuildDeepOperation {
  variables: VariableDefinitionNode[]
  fragments: FragmentDefinitionNode[]
  selections: SelectionNode[]
}

const buildDeepOperation = (props: CoreProps) => (
  target: Typename,
  path: Fieldname[],
  level = 0,
): BuildDeepOperation => {
  const [head, ...tail] = path

  const vtx = props.graph.get(target)

  if (!vtx) {
    throw Error('Invalid operation target')
  }

  const edgeData = vtx.weightsMap!.get(head)
  const nextTarget = vtx.edgesMap!.get(head)

  if (!edgeData || !nextTarget) {
    throw Error('Invalid operation path')
  }

  const argSuffix = level > 0 ? '' + level : ''

  const { variables, args } = (edgeData.arguments || []).reduce(
    (acc, input) => {
      acc.args.push(createOperationArgument(argSuffix)(input))
      acc.variables.push(createOperationVariable(argSuffix)(input))
      return acc
    },
    {
      variables: [] as VariableDefinitionNode[],
      args: [] as ArgumentNode[],
    },
  )

  // base
  if (tail.length === 0) {
    const { selections, fragmentnames } = buildSelection(props)([
      [head, nextTarget],
    ])

    const selection = {
      // this should always be field on the top level
      ...(selections[0] as FieldNode),
      // inject args
      arguments: args,
    }

    return {
      fragments: retriveCacheFragments(props)(fragmentnames),
      variables,
      selections: [selection],
    }
  } else {
    // not a base
    const {
      selections,
      fragments,
      variables: deepVariables,
    } = buildDeepOperation(props)(nextTarget, tail, level + 1)

    const nextSelection: FieldNode = createField({
      fieldname: head,
      arguments: args,
      selections,
    })

    return {
      variables: [...variables, ...deepVariables],
      fragments,
      selections: [nextSelection],
    }
  }
}

export const determineOperationType = (props: CoreProps) => (
  head: Fieldname,
) => {
  const query = props.graph.get('Query')
  if (query && query.edgesMap && query.edgesMap.has(head)) {
    return { vtx: query, type: 'query' as OperationTypeNode }
  }

  const mutation = props.graph.get('Mutation')
  if (mutation && mutation.edgesMap && mutation.edgesMap.has(head)) {
    return { vtx: query, type: 'mutation' as OperationTypeNode }
  }

  const subscription = props.graph.get('subscription')
  if (
    subscription &&
    subscription.edgesMap &&
    subscription.edgesMap.has(head)
  ) {
    return { vtx: query, type: 'subscription' as OperationTypeNode }
  }

  return { vtx: undefined, type: undefined }
}
