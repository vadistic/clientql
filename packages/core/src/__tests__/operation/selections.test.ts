import { print } from 'graphql'
import { createField, createFragment, wrapDocument } from '../../ast'
import { buildSelections, retriveFragmentsFromCache } from '../../operation'
import { complexProps, prismaProps, vatsProps } from '../fixture'

describe('selections', () => {
  it('prisma props > handle not very nested', () => {
    const selectionFn = buildSelections(prismaProps)

    const { selections, dependencies } = selectionFn('Post')

    const frag = createFragment({
      fragmentname: 'PostFragment',
      selections: selections!,
      condition: 'Post',
    })
    const fragments = retriveFragmentsFromCache(prismaProps)(dependencies)

    expect(print(wrapDocument(frag, ...fragments))).toMatchSnapshot()
  })

  it('prisma props > edges', () => {
    const selectionFn = buildSelections(prismaProps)

    const { selections, dependencies } = selectionFn('BoardEdge')

    const frag = createFragment({
      fragmentname: 'BoardedgeFragment',
      selections: selections!,
      condition: 'BoardEdge',
    })

    const fragments = retriveFragmentsFromCache(prismaProps)(dependencies)

    expect(print(wrapDocument(frag, ...fragments))).toMatchSnapshot()
  })

  it('complex props > selection on interface', () => {
    const selectionFn = buildSelections(complexProps)

    const { selections, dependencies } = selectionFn('Event')

    const fragments = retriveFragmentsFromCache(complexProps)(dependencies)

    const frag = createFragment({
      fragmentname: 'EventFragment',
      condition: 'Event',
      selections: [
        createField({
          fieldname: 'event',
          selections,
        }),
      ],
    })

    expect(print(wrapDocument(frag, ...fragments))).toMatchSnapshot()
  })

  it('vats props > ensure no empty flat fragments are generated', () => {
    const slectionFn = buildSelections(vatsProps)

    const { dependencies } = slectionFn('ApplicationConnection')

    const fragments = retriveFragmentsFromCache(vatsProps)(dependencies)

    expect(
      fragments.filter(frag => frag.selectionSet.selections.length === 0)
        .length,
    ).toBe(0)

    expect(fragments.map(print).join('\n\n')).toMatchSnapshot()
  })
})
