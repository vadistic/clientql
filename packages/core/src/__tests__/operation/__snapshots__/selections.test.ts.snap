// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selections complex props > selection on interface 1`] = `
"fragment EventFragment on Event {
  event {
    ...EventFlat
    venue {
      ...VenueFlat
    }
    ... on Concert {
      ...ConcertFlat
      venue {
        ...VenueFlat
      }
      previousVenues {
        ...VenueFlat
      }
      performingBand {
        ...PerformerFlat
      }
    }
    ... on Festival {
      ...FestivalFlat
      venue {
        ...VenueFlat
      }
      performers {
        ...PerformerFlat
      }
    }
    ... on Conference {
      ...ConferenceFlat
      venue {
        ...VenueFlat
      }
      speakers {
        ...SpeakerFlat
      }
    }
  }
}

fragment EventFlat on Event {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
}

fragment VenueFlat on Venue {
  id
  name
  address
  maxOccupancy
}

fragment ConcertFlat on Concert {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
}

fragment PerformerFlat on Performer {
  id
  name
  contact
  speciality
  fee
}

fragment FestivalFlat on Festival {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
}

fragment ConferenceFlat on Conference {
  id
  name
  startsAt
  endsAt
  minAgeRestriction
  workshops
}

fragment SpeakerFlat on Speaker {
  id
  name
  contact
}
"
`;

exports[`selections prisma props > edges 1`] = `
"fragment BoardedgeFragment on BoardEdge {
  ...BoardEdgeFlat
  node {
    ...BoardFlat
    admins {
      ...UserFlat
      posts {
        ...PostFlat
      }
    }
    members {
      ...UserFlat
      posts {
        ...PostFlat
      }
    }
    threads {
      ...ThreadFlat
      author {
        ...UserFlat
      }
      entry {
        ...PostFlat
      }
      replies {
        ...PostFlat
      }
    }
  }
}

fragment BoardEdgeFlat on BoardEdge {
  cursor
}

fragment BoardFlat on Board {
  id
  createdAt
}

fragment UserFlat on User {
  id
  nationality
  email
  name
}

fragment PostFlat on Post {
  id
  content
  createdAt
  published
  title
  updatedAt
}

fragment ThreadFlat on Thread {
  id
  createdAt
  name
}
"
`;

exports[`selections prisma props > handle not very nested 1`] = `
"fragment PostFragment on Post {
  ...PostFlat
  author {
    ...UserFlat
  }
}

fragment PostFlat on Post {
  id
  content
  createdAt
  published
  title
  updatedAt
}

fragment UserFlat on User {
  id
  nationality
  email
  name
}
"
`;

exports[`selections vats props > ensure no empty flat fragments are generated 1`] = `
"fragment PageInfoFlat on PageInfo {
  hasNextPage
  hasPreviousPage
  startCursor
  endCursor
}

fragment ApplicationEdgeFlat on ApplicationEdge {
  cursor
}

fragment ApplicationFlat on Application {
  createdAt
  id
  updatedAt
  type
}

fragment DisqualificationInstanceFlat on DisqualificationInstance {
  id
  createdAt
  updatedAt
  content
}

fragment StageFlat on Stage {
  id
  createdAt
  updatedAt
  name
  description
  type
  index
}

fragment ReviewInstanceFlat on ReviewInstance {
  id
  createdAt
  updatedAt
  rating
  content
}

fragment JobFlat on Job {
  id
  createdAt
  updatedAt
  type
  department
  name
  excerpt
  companyDescription
  description
  requirements
}

fragment CandidateFlat on Candidate {
  id
  createdAt
  updatedAt
  firstName
  lastName
  emails
  phones
  links
  company
  headline
  position
  resumesString
  coverLettersString
}

fragment AggregateApplicationFlat on AggregateApplication {
  count
}"
`;

exports[`selections vats props > ensure no empty fragments are generated 1`] = `
"fragment PageInfoFlat on PageInfo {
  hasNextPage
  hasPreviousPage
  startCursor
  endCursor
}

fragment ApplicationEdgeFlat on ApplicationEdge {
  cursor
}

fragment ApplicationFlat on Application {
  createdAt
  id
  updatedAt
  type
}

fragment DisqualificationInstanceFlat on DisqualificationInstance {
  id
  createdAt
  updatedAt
  content
}

fragment StageFlat on Stage {
  id
  createdAt
  updatedAt
  name
  description
  type
  index
}

fragment ReviewInstanceFlat on ReviewInstance {
  id
  createdAt
  updatedAt
  rating
  content
}

fragment JobFlat on Job {
  id
  createdAt
  updatedAt
  type
  department
  name
  excerpt
  companyDescription
  description
  requirements
}

fragment CandidateFlat on Candidate {
  id
  createdAt
  updatedAt
  firstName
  lastName
  emails
  phones
  links
  company
  headline
  position
  resumesString
  coverLettersString
}

fragment AggregateApplicationFlat on AggregateApplication {
  count
}"
`;
